doc/developer/design/20240610_unified_compute_introspection.md

```diff
@@ -0,0 +1,134 @@
+- ... can be retrieved successfully regardless of the health of any of the exporting replicas.
+
+## Out of Scope
+
+- Providing freshness of unified introspection data.
+  If a replica is unresponsive, its retrievable introspection data might be stale.
+
+## Solution Proposal
+
+This design describes an MVP implementation for Unified Compute Introspection that can provide value quickly.
+
+### Overview
+
+- The compute controller defines a set of _introspection subscribes_ that continuously read from introspection indexes, optionally transforming (filter/aggregate/join/...) the read data.
+- The catalog defines a set of `BuiltinSource`s and associated `IntrospectionType`s, one for each introspection subscribe.
+- Whenever a replica connection is established, the compute controller installs all defined introspection subscribes on the replica, as replica-targeted subscribes.
+- Whenever a batch of updates arrives for an introspection subscribe, the compute controller (a) prepends the replica ID to all updates and (b) sends the thus enriched updates to the storage controller, tagged with the corresponding `IntrospectionType`.
+- The storage controller's `CollectionManager` appends the introspection updates to the collection associated with the given `IntrospectionType`.
+- Whenever a replica disconnects, the compute controller cancels its introspection subscribes and instructs the storage controller to delete all introspection data previously emitted for this replica.
+- The storage controller's `CollectionManager` reads back, for each relevant `IntrospectionType`, the current contents of the storage collection, filters out updates belonging to the disconnected replica, and appends retractions for those to the storage collection.
+
+The `BuiltinSource`s fed by `IntrospectionType`s associated with introspection subscribes are unified introspection relations.
+They contain introspection updates for all live replicas and can be queried from any cluster.
+
+### Definition of a Unified Introspection Relation
+
+To create a new unified introspection relation, one has to define three parts:
+
+1. In the compute controller, a `DataflowDescription` that defines the introspection subscribe.
+2. In the storage controller, an `IntrospectionType` that defines the storage-managed collection into which updates from the introspection subscribe are written.
+3. In the catalog, a `BuiltinSource` that makes the storage-managed collection available to SQL queries.
+
+(2) and (3) are required for any `BuiltinSource` defined in the system. Only (1) is specific to this design.
+
+The `DataflowDescription`, since it is defined in the compute controller without access to the catalog or the optimizer, must be written using LIR constructs and can only reference introspection indexes, not builtin sources or views that may be defined in the catalog.
+These are limitations that seem reasonable for an MVP.
```

Comment:
Agree that that's OK for an MVP, but unlikely it's what we want for the final implementation. Expressing complicated queries for example with reductions is increasingly hard to do manually.

Reply:
A different take would be that all computations on introspection data should be done in the form of logging dataflows because they're running on all replicas and could have resource impact. If we restrict the plans to essentially be subscribes to existing arrangements, this would be the lowest cost possible because it allows us to hand-optimize the dataflows.

Reply:
That's true, but I doubt that handwriting a dataflow is any easier than writing the LIR definition (which you can avoid by adding a `dbg!(dataflow)` in the right place) :)

If we do this as a cost optimization, we'll want to avoid creating new indexes, so we'd have to introduce the concept of a "logging subscribe", i.e. an implicit subscribe that's installed on the replica at `CreateInstance` time. Seems feasible but not something we should consider if we don't observe a need. In contrast to the logging dataflows the introspection subscribes proposed in this design will show up in introspection, so we'll be able to see when they use non-negligible resources.